src/Servers/Kestrel/Core/src/Internal/Http2/HPack/Http2HPackEncoder.cs

```diff
@@ -0,0 +1,471 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Net.Http.HPack;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.HPack
+{
+    internal class Http2HPackEncoder
+    {
+        private static readonly int StatusHeaderSize = H2StaticTable.Get(H2StaticTable.Status200).Length;
+
+        // Internal for testing
+        internal readonly HPackHeaderEntry Head;
+        internal HPackHeaderEntry Removed;
+
+        private readonly IHeaderSensitivityDetector _sensitivityDetector;
+        private readonly HPackHeaderEntry[] _headerBuckets;
+        private readonly byte _hashMask;
+        private uint _headerTableSize;
+        private uint _maxHeaderTableSize;
+        private uint _maxHeaderListSize;
+        private bool _pendingTableSizeUpdate;
+
+        public bool IgnoreMaxHeaderListSize => _maxHeaderListSize == Http2PeerSettings.DefaultMaxHeaderListSize;
+
+        public Http2HPackEncoder(uint maxHeaderTableSize = Http2PeerSettings.DefaultHeaderTableSize, IHeaderSensitivityDetector sensitivityDetector = null)
+        {
+            _maxHeaderTableSize = maxHeaderTableSize;
+            _maxHeaderListSize = Http2PeerSettings.DefaultMaxHeaderListSize;
+            Head = new HPackHeaderEntry();
+            Head.Initialize(-1, string.Empty, string.Empty, int.MaxValue, null);
+            _headerBuckets = new HPackHeaderEntry[16];
```

Comment:
Any particular reason behind using 16 buckets?

Reply:
Other implementations allowed this value to be customized at a low-level. I didn't think we needed that level of control so I defaulted it. 16 was used in at least one other implementation (and it lines up with HashMap's default) so I used it.